### PR TITLE
Mark/Evoke transition system and lots of other goodies

### PR DESCRIPTION
**Pull request by rahul1980**
1. Modify the transition system by adding a MARK action to denote the beginning of a multi-token span. A MARK is paired with a subsequent EVOKE(type=t) action that evokes a span from the marked token to the current token. Marked tokens are pushed on a stack, which automatically gives a nested-semantics of most-nested-span-first.
  Meanwhile single token spans continue to be generated via EVOKE(length=1, type=t) actions, and the REFER action is unchanged. MARK actions are accompanied by four new features: binned distance from the current token to the top of the mark stack, the activation when the topmost marked token was pushed, the LR and RL LSTM vectors of the token where the MARK was output.  Splitting multi-token EVOKEs into MARK+EVOKE reduces the learning burden, and also gets rid of various EVOKE(len=L, type=T) actions where L>1.  This PR also increases the history feature's size from 4 to 5 to account for
the general increase in transition sequence lengths.
  Overall this boosts SPAN F1 from 92.5 -> 92.8, ROLE F1 from 72.2 -> 72.5, SLOT F1 from 82.8 -> 83.2.

2. This PR also simplifies the ParserState class by using Document's span-indexing code to judge when a span can be legally evoked. Doing this allows us to (a) get rid of the Nesting structure, (b) directly deal with frame handles instead of going via indices and integer handles, (c) get rid of the 'frames' array and solely deal with the attention buffer, (d) get rid of AddParseToDocument(), since frames/roles are eagerly added to the document as soon as they are generated.

3. Support for reading the flow file in the Python setup. This allows us to read the flow and perform inference. This PR adds a parse.py tool parallel to the parse.cc tool. Reading the flow file requires us to only read some simple fields for the Spec class, and all the big ticket items (e.g. commons, actions, cascade) are read off the commons blob.

4. Support for tracing. This PR adds adding the tracing information as a frame to the document, in both Python and C++. It is implemented as pay-as-you-go and when disabled, it doesn't cause any performance penalties (i.e. no timing differences). The tracing information has already proved invaluable whilst ensuring parity of the MARK/EVOKE C++ runtime vs Python. 
   The tracing file in Python (trace.py) can also be used as a standalone script that compares two recordios of documents with tracing information, and checks for equality of traces.  Right now it checks for lstm feature indices, ff features, and predicted & final actions. This was very useful in tracking down a few corner cases where the lexical features were different in Python vs C++.

5.  Minor changes:
- The transition generator has also been simplified a bit, given that parts of it had to be rewritten to accommodate the new MARK/EVOKE mechanism.
- We can now supply an output recordio file in parse.cc
- Remove the 'small' flag in Spec, since we can use the tracing utility to debug models now.
- Fix two lexical feature disparities: 
   (a) The C++ runtime was marking backquotes ("`") and carets ("^") as punctuation because CHARCAT_MODIFIER_SYMBOL was in the punctuation mask. I have removed it and now it behaves the same as in Python.
  (b) Fixed a missing encode('utf-8') call when I was generating suffixes that were unicode characters (there were only two such cases in the dev corpus).

### CODE REVIEW

**Comment by rahul1980** [comment]
Sorry about that. The c++ transition generator is now out of sync and
unused so we can delete it and ditto for the action table generator. I will
add that to this PR.

File: sling/nlp/document/lexical-encoder.cc [hunk 1]
```diff
@@ -313,6 +313,34 @@ int LexicalFeatures::InitWordEmbeddings(const string &filename) {
...
 };
 
+namespace {
+
+// Adds to 'trace', 'num' feature values located at 'feature' in 'data'.
+// Ignores -1, i.e. invalid lexical feature values.
+void AddTrace(int token, const string &name, int num,
+  myelin::Instance &data, myelin::Tensor *feature, Trace *trace) {
+  if (feature == nullptr) return;
+  const auto *location = data.Get<int>(feature);
```
Comment (ringgaard) [inline sling/nlp/document/lexical-encoder.cc]: What about const int * instead of auto? The name 'location' is also a bit confusing to me.

**Comment by rahul1980** [comment]
Michael, I have incorporated your changes for tracing lexical features in the lexical encoder. PTAL.

File: sling/nlp/parser/parser.cc [hunk 2]
```diff
@@ -116,11 +152,12 @@ void Parser::InitFF(const string &name, FF *ff) {
   ff->hidden = GetParam(name + "/hidden");
 }
 
-void Parser::Parse(Document *document) const {
+void Parser::Parse(Document *document, bool trace) const {
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/parser.cc]: Good idea, done.

File: sling/nlp/parser/parser.h [hunk 3]
```diff
@@ -72,17 +74,24 @@ class Parser {
...
     myelin::Tensor *labeled_roles_feature;    // labeled roles feature
 
+    int mark_depth = 0;                       // mark stack depth to use
     int attention_depth = 0;                  // number of attention features
     int history_size = 0;                     // number of history features
     int out_roles_size = 0;                   // max number of out roles
     int in_roles_size = 0;                    // max number of in roles
     int labeled_roles_size = 0;               // max number of unlabeled roles
     int unlabeled_roles_size = 0;             // max number of labeled roles
+    std::vector<int> mark_distance_bins;      // distance bins for mark tokens
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/parser.h]: Done.

File: sling/nlp/parser/parser.h [hunk 4]
```diff
@@ -165,8 +183,14 @@ class ParserInstance {
   CascadeInstance cascade_ = nullptr;
 
   // Frame creation and focus steps.
-  std::vector<int> create_step_;
-  std::vector<int> focus_step_;
+  HandleMap<int> create_step_;
+  HandleMap<int> focus_step_;
+
+  // (Token, step index) pushed by MARK actions.
+  std::vector<std::pair<int, int>> marks_;
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/parser.h]: Done.

File: sling/nlp/parser/trainer/train_util.py [hunk 5]
```diff
@@ -147,3 +156,30 @@ def setup_training_flags(flags):
...
+  flags.define('--parser',
+               help='Parser flow file',
+               default="",
+               type=str,
+               metavar='FLOW')
+  flags.define('--input',
+               help='Path to the input recordio file',
+               default="",
+               type=str,
+               metavar='INPUT_RECORDIO')
```
[the remark this answers is not in the document]
Comment (rahul1980) [reply sling/nlp/parser/trainer/train_util.py]: Done.

**Comment by rahul1980** [comment]
Thanks Michael, I have incorporated your suggestions. PTAL.

**Pull request merged**